File: packages/sdk-python/src/isl/results.py
"""
ISL Results - Typed result classes for API operations.

This module provides sealed-class-like result types using Python's
structural pattern matching (3.10+).

Example:
    >>> result = await client.users.create_user(input_data)
    >>> match result:
    ...     case CreateUserResult.Success(user=user):
    ...         print(f"Created: {user.id}")
    ...     case CreateUserResult.DuplicateEmail():
    ...         print("Email exists")
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, TypeVar, Callable, Any

from isl.models import User


T = TypeVar("T")


# =============================================================================
# Base Result Type
# =============================================================================


@dataclass(frozen=True, slots=True)
class Result(Generic[T]):
    """
    Base result type for all operations.
    
    Provides common helper methods for working with results.
    """
    
    @property
    def is_success(self) -> bool:
        """Check if result is successful."""
        return False
    
    @property
    def is_error(self) -> bool:
        """Check if result is an error."""
        return not self.is_success
    
    @property
    def value(self) -> T:
        """Get the success value. Raises if error."""
        raise NotImplementedError
    
    def unwrap(self) -> T:
        """
        Get the success value or raise an exception.
        
        Raises:
            ValueError: If result is an error
        """
        if self.is_success:
            return self.value
        raise ValueError(f"Cannot unwrap error result: {self}")
    
    def unwrap_or(self, default: T) -> T:
        """Get the success value or return a default."""
        return self.value if self.is_success else default
    
    def unwrap_or_else(self, default_fn: Callable[[], T]) -> T:
        """Get the success value or compute a default."""
        return self.value if self.is_success else default_fn()
    
    def map(self, fn: Callable[[T], Any]) -> Any:
        """Map the success value."""
        return fn(self.value) if self.is_success else None
    
    def on_success(self, fn: Callable[[T], None]) -> Result[T]:
        """Execute a function if successful."""
        if self.is_success:
            fn(self.value)
        return self
    
    def on_error(self, fn: Callable[[Result[T]], None]) -> Result[T]:
        """Execute a function if error."""
        if self.is_error:
            fn(self)
        return self


# =============================================================================
# CreateUser Results
# =============================================================================


class CreateUserResult:
    """
    Result type for create_user operation.
    
    Variants:
    - Success: User created successfully
    - DuplicateEmail: Email already exists
    - DuplicateUsername: Username already taken
    - InvalidInput: Input validation failed
    - RateLimited: Too many requests
    - ServerError: Server error
    - NetworkError: Network error
    """
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[User]):
        """Successful user creation."""
        user: User
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> User:
            return self.user
    
    @dataclass(frozen=True, slots=True)
    class DuplicateEmail(Result[User]):
        """Email already exists."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class DuplicateUsername(Result[User]):
        """Username already taken."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class InvalidInput(Result[User]):
        """Input validation failed."""
        message: str
        field: str | None = None
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class RateLimited(Result[User]):
        """Rate limit exceeded."""
        retry_after: int
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[User]):
        """Server error."""
        message: str
        code: str | None = None
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[User]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# =============================================================================
# GetUser Results
# =============================================================================


class GetUserResult:
    """Result type for get_user operation."""
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[User]):
        """User found."""
        user: User
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> User:
            return self.user
    
    @dataclass(frozen=True, slots=True)
    class NotFound(Result[User]):
        """User not found."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Unauthorized(Result[User]):
        """Not authorized."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[User]):
        """Server error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[User]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# =============================================================================
# UpdateUser Results
# =============================================================================


class UpdateUserResult:
    """Result type for update_user operation."""
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[User]):
        """User updated successfully."""
        user: User
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> User:
            return self.user
    
    @dataclass(frozen=True, slots=True)
    class NotFound(Result[User]):
        """User not found."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Unauthorized(Result[User]):
        """Not authorized."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Forbidden(Result[User]):
        """Operation forbidden."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class InvalidInput(Result[User]):
        """Input validation failed."""
        message: str
        field: str | None = None
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Conflict(Result[User]):
        """Conflict error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class RateLimited(Result[User]):
        """Rate limit exceeded."""
        retry_after: int
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[User]):
        """Server error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[User]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# =============================================================================
# DeleteUser Results
# =============================================================================


class DeleteUserResult:
    """Result type for delete_user operation."""
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[None]):
        """User deleted successfully."""
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> None:
            return None
    
    @dataclass(frozen=True, slots=True)
    class NotFound(Result[None]):
        """User not found."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Unauthorized(Result[None]):
        """Not authorized."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class Forbidden(Result[None]):
        """Operation forbidden."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[None]):
        """Server error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[None]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# =============================================================================
# ListUsers Results
# =============================================================================


class ListUsersResult:
    """Result type for list_users operation."""
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[list[User]]):
        """Users listed successfully."""
        users: list[User]
        next_page_token: str | None = None
        total_count: int | None = None
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> list[User]:
            return self.users
        
        @property
        def has_more(self) -> bool:
            return self.next_page_token is not None
    
    @dataclass(frozen=True, slots=True)
    class Unauthorized(Result[list[User]]):
        """Not authorized."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class InvalidInput(Result[list[User]]):
        """Input validation failed."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[list[User]]):
        """Server error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[list[User]]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# =============================================================================
# SearchUsers Results
# =============================================================================


class SearchUsersResult:
    """Result type for search_users operation."""
    
    @dataclass(frozen=True, slots=True)
    class Success(Result[list[User]]):
        """Search completed successfully."""
        users: list[User]
        next_page_token: str | None = None
        total_count: int | None = None
        
        @property
        def is_success(self) -> bool:
            return True
        
        @property
        def value(self) -> list[User]:
            return self.users
        
        @property
        def has_more(self) -> bool:
            return self.next_page_token is not None
    
    @dataclass(frozen=True, slots=True)
    class Unauthorized(Result[list[User]]):
        """Not authorized."""
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class InvalidInput(Result[list[User]]):
        """Input validation failed."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class ServerError(Result[list[User]]):
        """Server error."""
        message: str
        
        @property
        def is_success(self) -> bool:
            return False
    
    @dataclass(frozen=True, slots=True)
    class NetworkError(Result[list[User]]):
        """Network error."""
        cause: Exception
        
        @property
        def is_success(self) -> bool:
            return False


# Type aliases for convenience
CreateUserResultType = (
    CreateUserResult.Success
    | CreateUserResult.DuplicateEmail
    | CreateUserResult.DuplicateUsername
    | CreateUserResult.InvalidInput
    | CreateUserResult.RateLimited
    | CreateUserResult.ServerError
    | CreateUserResult.NetworkError
)

GetUserResultType = (
    GetUserResult.Success
    | GetUserResult.NotFound
    | GetUserResult.Unauthorized
    | GetUserResult.ServerError
    | GetUserResult.NetworkError
)

UpdateUserResultType = (
    UpdateUserResult.Success
    | UpdateUserResult.NotFound
    | UpdateUserResult.Unauthorized
    | UpdateUserResult.Forbidden
    | UpdateUserResult.InvalidInput
    | UpdateUserResult.Conflict
    | UpdateUserResult.RateLimited
    | UpdateUserResult.ServerError
    | UpdateUserResult.NetworkError
)

DeleteUserResultType = (
    DeleteUserResult.Success
    | DeleteUserResult.NotFound
    | DeleteUserResult.Unauthorized
    | DeleteUserResult.Forbidden
    | DeleteUserResult.ServerError
    | DeleteUserResult.NetworkError
)

ListUsersResultType = (
    ListUsersResult.Success
    | ListUsersResult.Unauthorized
    | ListUsersResult.InvalidInput
    | ListUsersResult.ServerError
    | ListUsersResult.NetworkError
)

SearchUsersResultType = (
    SearchUsersResult.Success
    | SearchUsersResult.Unauthorized
    | SearchUsersResult.InvalidInput
    | SearchUsersResult.ServerError
    | SearchUsersResult.NetworkError
)

File: packages/sdk-typescript/src/results.ts
/**
 * ISL Results - Typed result types for API operations.
 */

import type { User } from './models';

// =============================================================================
// Base Result Type
// =============================================================================

/**
 * Discriminated union result type
 */
export type Result<T, E> =
  | { readonly ok: true; readonly data: T }
  | { readonly ok: false; readonly error: E };

/**
 * Create a success result
 */
export function ok<T>(data: T): { ok: true; data: T } {
  return { ok: true, data };
}

/**
 * Create an error result
 */
export function err<E>(error: E): { ok: false; error: E } {
  return { ok: false, error };
}

/**
 * Result helper methods
 */
export const ResultHelpers = {
  /**
   * Unwrap a result or throw
   */
  unwrap<T, E>(result: Result<T, E>): T {
    if (result.ok) return result.data;
    throw new Error(`Cannot unwrap error result: ${JSON.stringify(result.error)}`);
  },

  /**
   * Unwrap with default value
   */
  unwrapOr<T, E>(result: Result<T, E>, defaultValue: T): T {
    return result.ok ? result.data : defaultValue;
  },

  /**
   * Unwrap with default factory
   */
  unwrapOrElse<T, E>(result: Result<T, E>, fn: () => T): T {
    return result.ok ? result.data : fn();
  },

  /**
   * Map success value
   */
  map<T, E, U>(result: Result<T, E>, fn: (value: T) => U): Result<U, E> {
    return result.ok ? ok(fn(result.data)) : result;
  },

  /**
   * Map error value
   */
  mapError<T, E, F>(result: Result<T, E>, fn: (error: E) => F): Result<T, F> {
    return result.ok ? result : err(fn(result.error));
  },

  /**
   * Flat map result
   */
  flatMap<T, E, U>(result: Result<T, E>, fn: (value: T) => Result<U, E>): Result<U, E> {
    return result.ok ? fn(result.data) : result;
  },
};

// =============================================================================
// Error Types
// =============================================================================

/**
 * Base error structure
 */
export interface BaseError {
  readonly code: string;
  readonly message: string;
}

/**
 * Create user errors
 */
export type CreateUserError =
  | { readonly code: 'DUPLICATE_EMAIL' }
  | { readonly code: 'DUPLICATE_USERNAME' }
  | { readonly code: 'INVALID_INPUT'; readonly message: string; readonly field?: string }
  | { readonly code: 'RATE_LIMITED'; readonly retryAfter: number }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

/**
 * Get user errors
 */
export type GetUserError =
  | { readonly code: 'NOT_FOUND' }
  | { readonly code: 'UNAUTHORIZED' }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

/**
 * Update user errors
 */
export type UpdateUserError =
  | { readonly code: 'NOT_FOUND' }
  | { readonly code: 'UNAUTHORIZED' }
  | { readonly code: 'FORBIDDEN' }
  | { readonly code: 'INVALID_INPUT'; readonly message: string; readonly field?: string }
  | { readonly code: 'CONFLICT'; readonly message: string }
  | { readonly code: 'RATE_LIMITED'; readonly retryAfter: number }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

/**
 * Delete user errors
 */
export type DeleteUserError =
  | { readonly code: 'NOT_FOUND' }
  | { readonly code: 'UNAUTHORIZED' }
  | { readonly code: 'FORBIDDEN' }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

/**
 * List users errors
 */
export type ListUsersError =
  | { readonly code: 'UNAUTHORIZED' }
  | { readonly code: 'INVALID_INPUT'; readonly message: string }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

/**
 * Search users errors
 */
export type SearchUsersError =
  | { readonly code: 'UNAUTHORIZED' }
  | { readonly code: 'INVALID_INPUT'; readonly message: string }
  | { readonly code: 'SERVER_ERROR'; readonly message: string }
  | { readonly code: 'NETWORK_ERROR'; readonly cause?: Error };

// =============================================================================
// Result Types
// =============================================================================

/**
 * Create user result
 */
export type CreateUserResult = Result<User, CreateUserError>;

/**
 * Get user result
 */
export type GetUserResult = Result<User, GetUserError>;

/**
 * Update user result
 */
export type UpdateUserResult = Result<User, UpdateUserError>;

/**
 * Delete user result
 */
export type DeleteUserResult = Result<void, DeleteUserError>;

/**
 * List users result
 */
export type ListUsersResult = Result<
  {
    readonly users: readonly User[];
    readonly nextPageToken?: string;
    readonly totalCount?: number;
  },
  ListUsersError
>;

/**
 * Search users result
 */
export type SearchUsersResult = Result<
  {
    readonly users: readonly User[];
    readonly nextPageToken?: string;
    readonly totalCount?: number;
  },
  SearchUsersError
>;

// =============================================================================
// Error Factories
// =============================================================================

export const CreateUserErrors = {
  duplicateEmail: (): CreateUserError => ({ code: 'DUPLICATE_EMAIL' }),
  duplicateUsername: (): CreateUserError => ({ code: 'DUPLICATE_USERNAME' }),
  invalidInput: (message: string, field?: string): CreateUserError => ({
    code: 'INVALID_INPUT',
    message,
    field,
  }),
  rateLimited: (retryAfter: number): CreateUserError => ({
    code: 'RATE_LIMITED',
    retryAfter,
  }),
  serverError: (message: string): CreateUserError => ({
    code: 'SERVER_ERROR',
    message,
  }),
  networkError: (cause?: Error): CreateUserError => ({
    code: 'NETWORK_ERROR',
    cause,
  }),
};

export const GetUserErrors = {
  notFound: (): GetUserError => ({ code: 'NOT_FOUND' }),
  unauthorized: (): GetUserError => ({ code: 'UNAUTHORIZED' }),
  serverError: (message: string): GetUserError => ({
    code: 'SERVER_ERROR',
    message,
  }),
  networkError: (cause?: Error): GetUserError => ({
    code: 'NETWORK_ERROR',
    cause,
  }),
};
